### ISSUE TITLE
Crash: Unable to write mutant diff correctly as hashes are different

### ISSUE THREAD

**Issue by jkrauska** [CodaProtocol/coda#2158]
Possible persistence bug encountered on testnet.

```
{"timestamp":"2019-04-05 00:42:17.523731Z","level":"Warn","source":{"module":"Coda","location":"File \"app/cli/src/coda.ml\", line 301, characters 60-67"},"message":"long async cycle 1.18616s","metadata":{"pid":9183}}
(monitor.ml.Error
 (Failure "Unable to write mutant diff correctly as hashes are different")
 ("Raised at file \"stdlib.ml\", line 33, characters 17-33"
  "Called from file \"src/deferred1.ml\", line 20, characters 40-45"
  "Called from file \"src/job_queue.ml\" (inlined), line 120, characters 2-5"
  "Called from file \"src/job_queue.ml\", line 159, characters 6-47"
  "Caught by monitor coda"))
```

Note: this was right after a long async cycle.

[mutant.tar.gz](https://github.com/CodaProtocol/coda/files/3045976/mutant.tar.gz)

**Comment by cmr** [comment]
Master test-unit--dev flaked with this failure.

backtrace: https://gist.github.com/cmr/c380353381d1c35bb754941551fccee9
ci job: https://circleci.com/gh/CodaProtocol/coda/90606?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by cmr** [comment]
We're seeing this intermittently still. @johnwu93 spent some time trying to get a reliable repro but hasn't found one yet. We have some ideas. It's next on his plate.

**Comment by jkrauska** [comment]
@nholland94 is currently refactoring external_transitions, which would complicate merging persistance changes related to this bug -- so effort is on hold....

**Comment by johnwu93** [comment]
I will start reworking on this after the refactoring is on a PR

**Comment by cmr** [comment]
We're seeing this again. And we still see it in unit tests sometimes.

**Comment by psteckler** [comment]
I was looking at nondeterminism in `bin_io` serialization as a possible source of this error, since the computed hashes depend on `Binable.to_string`. But I didn't find a definite source of the error.

I did verify that the TF persistence data being written to RocksDB was being retrieved intact.

The error occurs when the diff is an `Update_root`, and it's only the `scan_state` field that appears to differ. I tried making a copy of that field in the diff. The error still occurred.

@nholland94 says that the hashes are not being computed correctly, and he's doing refactoring. I'm pausing working on this, for now.
